### ISSUE TITLE
C# Async support

### ISSUE THREAD

**Issue by pluskal** [kaitai-io/kaitai_struct#640]
Have you considered to extend C# support with async?

Currently, `CSharpCompiler` generates parsers with synchronous API. It would be great to be able to use Kaitai for parsing data asynchronously, e.g., a custom network protocol parsing on the highly concurrent server. This would increase throughput and lower latency and memory footprint of such an application. Usage of the synchronous Kaitai parsers in asynchronous calls results in the depletion of ThreadPool.

After a brief study of Kaitai, I assume that the `KaitaiStream`, and `CSharpCompiler` are the candidates for modification. 

Because parses in Kaitai heavily use `Seek` that has no `Async` overload in BLC, the change would most probably require to wrap the `Stream` in `KaitaiStream` into custom implementation and introduce some buffering. Also, the `BinaryReader` used as a base of `KaitaiStream` does not provide async API so that custom implementation would be required as well.

Would you consider a PR containing such modifications?

**Comment by GreyCat** [comment]
Makes sense, all PRs are always very welcome. If I may suggest, given that this seems to be relatively substantial update, may be we could start with a design / plan / architecture description, ask everyone interested to review/discuss it, and proceed with the implementation.

Looks like it is the candidate for either a new target (~as a new language), with its own set of tests, its own async runtime, generated APIs, etc.

**Comment by pluskal** [comment]
Agreed! I need to dig a little bit into Kaitai to see what methods are required for the `KaitaiStream\KaitaiStruct` to have. 

Also, I agree that the new target would be better because the current C# implementation does parsing in the constructor. That cannot be done in async, so to have both sync and async parser in one class would require a breaking change to the parser API. But it is up to you to decide. If you are ok with a breaking change, it is common to have both sync and async, APIs in one C# class. 

I can prepare a prototype of C# async runtime, but I am afraid that I will not be able to do much with `CSharpCompiler` except a simple async modification.

**Comment by GreyCat** [comment]
> Also, I agree that the new target would be better because the current C# implementation does parsing in the constructor. That cannot be done in async, so to have both sync and async parser in one class would require a breaking change to the parser API.

Actually, constructor nowadays optionally calls `_read()` method, and all parsing happens in there.

Also, from the testing perspective, I believe we'll need to test *both* anyway, so at the very least we'll be having (may be "virtual") different targets, as we do currently with `cpp_stl_98` (which actually corresponds to `--target=cpp_stl --cpp-version=98`) vs `cpp_stl_11` (equals to `--target=cpp_stl --cpp-version=98`).

**Comment by pluskal** [comment]
Can you look at this messy sketch? I tried to add Async overloads for all required methods, and some cleanups. 

https://github.com/pluskal/kaitai_struct_csharp_runtime/pull/2

There are several possibilities for how to use Async parsing. In my opinion, it would be nice to have an opportunity to inherit `KaitaiStream` and handle some crucial operations that otherwise block async calls like `Position, Seek(), Length`. In my case, I'll be using most probably `System.IO.Pipelines`.  For this, it would be nice to be able to reimplement `KaitaiStream` completely, if needed, so I extracted `sync`, and `async` interfaces that describe what needs to be implemented. Generated parsers can you the interfaces instead of concrete type of `KaitaiStream`.

Some general thoughts beyond the scope of this issue:
- Change from net4.5 to .netstandard2.0. If you need to support net4.5, it can be done by dual targeting without Async. Update csproj. `kaitai_struct_runtime_csharp.nuspec` can be removed and package can be generated from info in new csproj by `dotnet pack`
- Split multi class definitions in one file to separate files
- Add sln file and test projects
- Perhaps submodule or guidance where to find up-to-date parser samples - I tried [formats-kaitai-io.github.io](https://github.com/kaitai-io/formats-kaitai-io.github.io) that were not compatible with `kaitai_struct_csharp_runtime` master
- Cleanup code a bit according to C# standard naming conventions - I tried some, but I did not touch public members so it should stay compatible. 

If you are interested in these changes, I can create separate issues and submit PRs.

**Comment by GreyCat** [comment]
Thanks for sharing this! Looks good to me, although I'd appreciate some feedback from our C# experts  — cc @koczkatamas, @LogicAndTrick, @indrora?

> For this, it would be nice to be able to reimplement KaitaiStream completely, if needed, so I extracted sync, and async interfaces that describe what needs to be implemented. Generated parsers can you the interfaces instead of concrete type of KaitaiStream.

Actually, the very similar approach is used in Java: [KaitaiStream](https://github.com/kaitai-io/kaitai_struct_java_runtime/blob/master/src/main/java/io/kaitai/struct/KaitaiStream.java) is a partial implementation of methods common to all underlying streams, which is then extended by [ByteBufferKaitaiStream](https://github.com/kaitai-io/kaitai_struct_java_runtime/blob/master/src/main/java/io/kaitai/struct/ByteBufferKaitaiStream.java) (based on ByteBuffer) and [RandomAccessFileKaitaiStream](https://github.com/kaitai-io/kaitai_struct_java_runtime/blob/master/src/main/java/io/kaitai/struct/RandomAccessFileKaitaiStream.java) (based on RandomAccessFile). It likely makes sense to extract pure interface too, in case somebody wants to reimplement that completely.

There are a bunch of static methods which are hosted inside KaitaiStream for sake of simplicity (like `ProcessXor`, etc). Given that we're expanding number of files in the repo anyway, it does not make sense to "hide" them all in a single class. They really don't have much to do with KaitaiStream per se, so we might want to extract them somewhere else (like KaitaiUtils)?

> Some general thoughts beyond the scope of this issue:

Let's go in the small steps and do these as separate PRs, adjusting codegen as needed along the way to not avoid breaking things radically ;)

> Change from net4.5 to .netstandard2.0. If you need to support net4.5, it can be done by dual targeting without Async.

I believe we do. .NET 4.5 is still a very big share of the potential users. Can you help with this dual targeting?

> `kaitai_struct_runtime_csharp.nuspec` can be removed and package can be generated from info in new csproj by `dotnet pack`

Ok, that's likely good news ;)

> Split multi class definitions in one file to separate files

Agreed. As far as I understand, modern C# style guides require that anyway.

> Add sln file and test projects

There is actually a [sln file in tests repo](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/spec/csharp/kaitai_struct_csharp_tests.sln). What do you mean by "test projects"?

> Perhaps submodule or guidance where to find up-to-date parser samples - I tried [formats-kaitai-io.github.io](https://github.com/kaitai-io/formats-kaitai-io.github.io) that were not compatible with `kaitai_struct_csharp_runtime` master

https://formats.kaitai.io/ currently target latest stable version, i.e. 0.8. It should be compatible with relevant tag in runtime repo — but, alas, [we forgot to assign one](https://github.com/kaitai-io/kaitai_struct_csharp_runtime/tags).

> Cleanup code a bit according to C# standard naming conventions - I tried some, but I did not touch public members so it should stay compatible.

I believe we can start by adding StyleCop with something like "Microsoft recommended managed rules" enabled by default on this repo.

**Comment by pluskal** [comment]
Thanks for the tips with Java KaitaiStream. I'll try to do it in a similar manner.

I started with the [multi-targeting](https://github.com/kaitai-io/kaitai_struct_csharp_runtime/pull/15) and move forward after your review. I agree with all your comments.

> There is actually a sln file in tests repo. What do you mean by "test projects"

I missed that you have a separate repo with tests. I was expecting them to be with the source. I need to look more closely into your build system.
